src/sys/system_v/shm.rs

```diff
@@ -0,0 +1,476 @@
+    marker::PhantomData,
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+    ptr,
+};
+
+use crate::Result;
+use crate::{errno::Errno, sys::stat::Mode};
+
+use libc::{self, c_int, c_void, key_t, shmid_ds};
+
+#[derive(Debug)]
+/// Safe wrapper to create and connect to a SystemV shared memory segment.
+///
+/// # Example
+///
+/// ```no_run
+/// # use nix::errno::Errno;
+/// # use nix::sys::system_v::shm::*;
+/// # use nix::sys::stat::Mode;
+/// #
+/// struct MyData(i64);
+///
+/// const MY_KEY: i32 = 1337;
+/// let mem_segment = Shm::<MyData>::create_and_connect(
+///     MY_KEY,
+///     Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+/// )?;
+/// let mut shared_memory = mem_segment.attach(ShmatFlag::empty())?;
+/// # Ok::<(), Errno>(())
+/// ```
+///
+pub struct Shm<T> {
+    id: c_int,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Shm<T> {
+    /// Attach to the current SystemV shared memory segment.
+    ///
+    /// To create a new segment, use [`Shm::create_and_connect`].\
+    /// If you need more customisation, use the unsafe version,
+    /// [`Shm::shmget`], with the key [`ShmgetFlag::IPC_CREAT`].
+    ///
+    /// Attaching a segment to a specific adress isn't supported. This is
+    /// because there is no way to create a void pointer on rust.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use nix::errno::Errno;
+    /// # use nix::sys::system_v::shm::*;
+    /// # use nix::sys::stat::Mode;
+    /// #
+    /// struct MyData(i64);
+    ///
+    /// const MY_KEY: i32 = 1337;
+    /// let mem_segment = Shm::<MyData>::create_and_connect(
+    ///     MY_KEY,
+    ///     Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+    /// )?;
+    /// let mut shared_memory = mem_segment.attach(ShmatFlag::empty())?;
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    pub fn attach(&self, shmat_flag: ShmatFlag) -> Result<SharedMemory<T>> {
+        unsafe {
+            Ok(SharedMemory::<T> {
+                id: self.id,
+                shm: ManuallyDrop::new(Box::from_raw(self.shmat(shmat_flag)?)),
+            })
+        }
+    }
+
+    /// Creates and returns a new System V shared memory segment identifier.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use nix::errno::Errno;
+    /// # use nix::sys::system_v::shm::*;
+    /// # use nix::sys::stat::Mode;
+    /// #
+    /// struct MyData(i64);
+    /// const MY_KEY: i32 = 1337;
+    ///
+    /// let mem_segment = Shm::<MyData>::create_and_connect(
+    ///     MY_KEY,
+    ///     Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+    /// )?;
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    pub fn create_and_connect(key: key_t, mode: Mode) -> Result<Self> {
+        let size = std::mem::size_of::<T>();
+        // This is the main difference between this function and [`Shm::shmget`]
+        // Because we are always creating a new segment, we can be sure that the size match
+        let shmget_flag = ShmgetFlag::IPC_CREAT | ShmgetFlag::IPC_EXCL;
```

Comment:
I do not understand what you mean here.
Why would `.attach()` use `shmget_flag`, when it only connect to an existing memory segment?

This here is guaranteed safe because we're 100% sure that the new memory segment either has the size of `T`, or if it fails to be created, return an error. If something more fine grained is needed, isn't `unsafe .shmget()` the function that fulfill this purpose?